File: wlr/SPX.Core/Input.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Input;

namespace SPX.Core
{
    public interface IInputInitializer
    {
        ICollection<CommandDefinition> GetCommands();
    }
    public class Contexts
    {
        public const int All = 0;
        public const int Nonfree = 1;
        public const int Free = 2;
        public const int Inventory = 3;
    }
    public class CommandDefinition
    {
        public CommandDefinition(int command, Keys key, Buttons button, int lockContext)
        {
            Command = command;
            Gamepad = button;
            Keyboard = key;
            LockContext = lockContext;
        }
        public int Command{get;set;}
        public Buttons Gamepad{get;set;}
        public Keys Keyboard{get;set;}
        public int LockContext { get; set; }
    }
    public class Input
    {        
        //Maps a playerId to a context
        private static readonly Dictionary<int, int> __contexts = 
            new Dictionary<int, int>()
            {
                {0, Contexts.Free},
                {1, Contexts.Free},
                {2, Contexts.Free},
                {3, Contexts.Free}
            };

        //Lists what commands are locked for a given player
        private static readonly List<CommandLock> __locks = new List<CommandLock>(); 

        //The commands that cannot be used by simply holding down the command's input depending on the given context
        private static readonly Dictionary<int, int> __lockOnPress = new Dictionary<int, int>();

        private static readonly Dictionary<int, Keys> _keyboardMapping = new Dictionary<int, Keys>();

        private static readonly Dictionary<int, Buttons> _gamePadMapping = new Dictionary<int, Buttons>();

        private static readonly List<PlayerIndex> _playerIndex = new List<PlayerIndex>()
                                                                      {
                                                                          PlayerIndex.One,
                                                                          PlayerIndex.Two,
                                                                          PlayerIndex.Three,
                                                                          PlayerIndex.Four
                                                                      };
        private static readonly Dictionary<int, bool> __inputs = new Dictionary<int, bool>();
        private static bool __isInputActive = false;
        private static bool __isDown = false;

        private static int __controllerCount = 0;

        public static int GetPlayerCount()
        {
            return __controllerCount;
        }

        public static void Setup(IInputInitializer initializer)
        {
            foreach (var command in initializer.GetCommands())
            {
                _keyboardMapping.Add(command.Command, command.Keyboard);
                _gamePadMapping.Add(command.Command, command.Gamepad);
                if (command.LockContext >= 0)
                {
                    __lockOnPress.Add(command.Command, command.LockContext);
                }
            }

            //Call an update to detect controllers that are connected
            Update();
        }

        private static bool IsDown(int command, int playerIndex)
        {
            return GamePad.GetState(_playerIndex[playerIndex]).IsButtonDown(_gamePadMapping[command])
            || 
            (playerIndex == 0 && Keyboard.GetState().IsKeyDown(_keyboardMapping[command]));
        }

        public static bool IsPressed(int command, int playerIndex,bool failIfLocked=true)
        {
            
            __isInputActive = IsDown(command,playerIndex);
            if (!__isInputActive && ShouldLock(command,playerIndex))
            {
                Unlock(command, playerIndex);
            }

            if (IsLocked(command, playerIndex) && failIfLocked)
            {
                return false;
            }

            if (__isInputActive && ShouldLock(command,playerIndex))
            {
                Lock(command,playerIndex);
            }

            return __isInputActive;
        }

        //If the key is marked to be locked on press and its lock context is currently inactive
        private static bool ShouldLock(int command,int playerIndex)
        {
            foreach(var key in __lockOnPress.Keys)
            {
                if (key == command)
                {
                    if (__lockOnPress[key] == __contexts[playerIndex] ||
                        (__lockOnPress[key] == Contexts.Nonfree && __contexts[playerIndex] != Contexts.Free) ||
                        __lockOnPress[key] == Contexts.All)
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        public static void SetContext(int context,int playerIndex)
        {
            __contexts[playerIndex] = context;
        }
        public static bool IsContext(int context,int playerIndex)
        {
            return __contexts[playerIndex] == context;
        }
        public static bool IsLocked(int command,int playerIndex)
        {
            foreach (var pair in __locks)
            {
                if (pair.Command == command && pair.PlayerIndex == playerIndex)
                {
                    return true;
                }
            }
            return false;
        }
        public static void Lock(int command,int playerIndex)
        {
            __locks.Add(new CommandLock(command,playerIndex));
        }
        public static void Unlock(int command, int playerIndex)
        {
            for(int ii = 0;ii<__locks.Count();ii++)
            {
                if(__locks[ii].Command==command && __locks[ii].PlayerIndex==playerIndex)
                {
                    __locks.Remove(__locks[ii]);
                    ii--;
                }
            }
        }

        private static readonly List<PlayerIndex> __playerIndices = new List<PlayerIndex>()
        {
            PlayerIndex.One,
            PlayerIndex.Two,
            PlayerIndex.Three,
            PlayerIndex.Four,
        };

        public static void Update()
        {
            //Remove command locks if the associated key/button isn't being pressed
            for(int ii = 0;ii<__locks.Count();ii++)
            {
                if (!IsDown(__locks[ii].Command, __locks[ii].PlayerIndex))
                {
                    __locks.Remove(__locks[ii]);
                    ii--;
                }
            }

            //Update the number of players currently connected via controller
            for(int ii = 0;ii<__playerIndices.Count();ii++)
            {
                __controllerCount += (GamePad.GetState(__playerIndices[ii]).IsConnected)?1:0;
                if (!__inputs.ContainsKey(ii))
                {
                    __inputs.Add(ii, GamePad.GetState(_playerIndex[ii]).IsConnected);
                } 
            }
            if (__controllerCount == 0)
            {
                __controllerCount = 1;
            }
        }
        private class CommandLock
        {
            public CommandLock(int command, int playerIndex)
            {
                Command = command;
                PlayerIndex = playerIndex;
            }
            public int Command;
            public int PlayerIndex { get; set; }
        }
    }
}

File: aigilas/src/sps/io/Input.java
package sps.io;

import aigilas.Aigilas;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.controllers.Controllers;
import sps.bridge.Command;
import sps.bridge.Context;
import sps.bridge.Contexts;
import sps.bridge.Sps;
import sps.core.Logger;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Input implements InputProvider {

    private static InputProvider instance = new Input();
    private static InputProvider falseInstance = new FalseInput();
    private static boolean disabled = false;
    // Lists what commands are locked for a given player
    private final List<CommandLock> __locks = new ArrayList<CommandLock>();
    private StateProvider provider;
    // $$$ FIXME (Integer -> PlayerId) Maps a playerId to a context
    private HashMap<Integer, Context> __contexts;
    private boolean __isInputActive = false;

    public static InputProvider get() {
        return isDisabled() ? falseInstance : instance;
    }

    public static void enable() {
        disabled = false;
    }

    public static void disable() {
        disabled = true;
    }

    public static boolean isDisabled() {
        return disabled;
    }

    @Override
    public void setup(StateProvider stateProvider) {
        __contexts = new HashMap<Integer, Context>();
        __contexts.put(0, Contexts.get(Sps.Contexts.Free));
        __contexts.put(1, Contexts.get(Sps.Contexts.Free));
        __contexts.put(2, Contexts.get(Sps.Contexts.Free));
        __contexts.put(3, Contexts.get(Sps.Contexts.Free));

        if (stateProvider == null) {
            provider = new DefaultStateProvider();
        }
        else {
            provider = stateProvider;
        }

        InputBindings.init();
    }

    @Override
    public boolean detectState(Command command, int playerIndex) {
        boolean debugInput = false;
        boolean gamepadActive = false;
        if (Controllers.getControllers().size > playerIndex) {
            gamepadActive = command.controllerInput().isActive(Controllers.getControllers().get(playerIndex));
        }
        boolean keyboardActive = (playerIndex == provider.getFirstPlayerIndex()) && Gdx.input.isKeyPressed(command.key().getKeyCode());

        if (debugInput && command.equals(Aigilas.Commands.MoveLeft)) {
            if (debugInput && (gamepadActive || keyboardActive)) {
                Logger.info("ACTIVE: " + command.name() + ", controller:" + gamepadActive + ", keyboard:" + keyboardActive);
            }
            else {
                Logger.info("INACTIVE: " + command.name() + ", controller:" + gamepadActive + ", keyboard:" + keyboardActive);
            }
        }

        return gamepadActive || keyboardActive;
    }

    private boolean isDown(Command command, int playerIndex) {
        return provider.isActive(command, playerIndex);
    }

    public boolean isActive(Command command, int playerIndex) {
        return isActive(command, playerIndex, true);
    }

    @Override
    public boolean isActive(Command command, int playerIndex, boolean failIfLocked) {
        __isInputActive = isDown(command, playerIndex);
        if (!__isInputActive && shouldLock(command, playerIndex)) {
            unlock(command, playerIndex);
        }

        if (isLocked(command, playerIndex) && failIfLocked) {
            return false;
        }

        if (__isInputActive && shouldLock(command, playerIndex)) {
            lock(command, playerIndex);
        }

        return __isInputActive;
    }

    // If the key is marked to be locked on press and its lock context is
    // currently inactive
    private boolean shouldLock(Command command, int playerIndex) {
        return command.Context == __contexts.get(playerIndex) || (command.Context == Contexts.get(Sps.Contexts.Non_Free) && __contexts.get(playerIndex) != Contexts.get(Sps.Contexts.Free) || command.Context == Contexts.get(Sps.Contexts.All));
    }

    @Override
    public void setContext(Context context, int playerIndex) {
        __contexts.put(playerIndex, context);
    }

    @Override
    public boolean isContext(Context context, int playerIndex) {
        return __contexts.get(playerIndex) == context;
    }

    @Override
    public boolean isLocked(Command command, int playerIndex) {
        for (CommandLock pair : __locks) {
            if (pair.Command == command && pair.PlayerIndex == playerIndex) {
                return true;
            }
        }
        return false;
    }

    @Override
    public void lock(Command command, int playerIndex) {
        __locks.add(new CommandLock(command, playerIndex));
    }

    @Override
    public void unlock(Command command, int playerIndex) {
        for (int ii = 0; ii < __locks.size(); ii++) {
            if (__locks.get(ii).Command == command && __locks.get(ii).PlayerIndex == playerIndex) {
                __locks.remove(__locks.get(ii));
                ii--;
            }
        }
    }

    @Override
    public void update() {
        // Remove command locks if the associated key/button isn't being pressed
        for (int ii = 0; ii < __locks.size(); ii++) {
            if (!isDown(__locks.get(ii).Command, __locks.get(ii).PlayerIndex)) {
                __locks.remove(__locks.get(ii));
                ii--;
            }
        }

        provider.pollLocalState();
    }
}
